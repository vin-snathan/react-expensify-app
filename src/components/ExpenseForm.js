import React, { Component } from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';

class ExpenseForm extends Component {
	constructor(props) {
		super(props);
		this.state = {description: props.description || '', amount: props.amount ? (props.amount/100).toString() : '', note: props.note || '', createdAt: moment(props.createdAt) || moment(), calendarFocused: false, error: ''};
		this.onDescriptionChange = this.onDescriptionChange.bind(this);
		this.onAmountChange = this.onAmountChange.bind(this);
		this.onNoteChange = this.onNoteChange.bind(this);
		this.onDateChange = this.onDateChange.bind(this);
		this.onFocusChange = this.onFocusChange.bind(this);
		this.onFormSubmit = this.onFormSubmit.bind(this);
	}

	onDescriptionChange(e) {
		const description = e.target.value;
		this.setState(() => ({description}))
	}

	onAmountChange(e) {
		const amount = e.target.value;
		if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
			this.setState(() => ({amount}));
		}
	}

	onNoteChange(e) {
		const note = e.target.value;
		this.setState(() => ({note}))
	}

	onDateChange(createdAt) {
		this.setState(() => ({createdAt}))
	}

	onFocusChange({focused}) {
		this.setState(() => ({calendarFocused: focused}));
	}

	onFormSubmit(e) {
		e.preventDefault();

		if(!this.state.description || !this.state.amount) {
			this.setState(() => ({error: 'Please provide description and amount'}))
		} else {
			this.setState(() => ({error: ''}));
			this.props.onSubmit({
				description: this.state.description,
				amount: parseFloat(this.state.amount, 10) * 100,
				createdAt: this.state.createdAt.valueOf(),
				note: this.state.note
			})
		}
	}

	render() {
		return (
			<div>
				{this.state.error && <p style={{color: 'red'}}> {this.state.error} </p>}
				<form onSubmit={this.onFormSubmit}>
					<input type='text' placeholder='Description' value={this.state.description} onChange={this.onDescriptionChange} autoFocus />
					<input type='text' placeholder='Amount' value={this.state.amount} onChange={this.onAmountChange} />
					<SingleDatePicker
					  date={this.state.createdAt} // momentPropTypes.momentObj or null
					  onDateChange={this.onDateChange} // PropTypes.func.isRequired
					  focused={this.state.calendarFocused} // PropTypes.bool
					  onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
					  isOutsideRange={() => false}
					  numberOfMonths={1}
					/>
					<textarea placeholder='Add a note for your expense (optional)' value={this.state.note} onChange={this.onNoteChange}></textarea>
					<button type='submit'> Add Expense </button>
				</form>
			</div>
		);
	}
} 

export default ExpenseForm;