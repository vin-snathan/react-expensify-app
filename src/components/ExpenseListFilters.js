import React, {Component} from 'react';
import {connect} from 'react-redux';
import {setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate} from '../actions/filters';
import moment from 'moment';
import {DateRangePicker} from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

export class ExpenseListFilters extends Component {
	constructor(props) {
		super(props);
		this.state = {focusedInput: null}
		this.handleSortBy = this.handleSortBy.bind(this);
		this.onFocusChange = this.onFocusChange.bind(this);
		this.onDatesChange = this.onDatesChange.bind(this);
		this.onTextChange = this.onTextChange.bind(this);
	}

	handleSortBy(e) {
		e.target.value === 'amount' ? this.props.sortByAmount() : this.props.sortByDate();
	}

	onDatesChange({startDate, endDate}) {
		this.props.setStartDate(startDate);
		this.props.setEndDate(endDate);
	}

	onTextChange(e) {
		this.props.setTextFilter(e.target.value);
	}

	onFocusChange(focusedInput) {
		this.setState(() => ({focusedInput}));
	}

	render() {
		return (
			<div>
				<input type='text' value={this.props.filters.text} name='textFilter' onChange={this.onTextChange} />

				<select value={this.props.filters.sortBy} onChange={this.handleSortBy}>
				  <option value='amount'>Amount</option>
				  <option value='date'>Date</option>
				</select>

				<DateRangePicker
				  startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
				  startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
				  endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
				  endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
				  onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
				  focusedInput={this.state.focusedInput} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
				  onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
				  numberOfMonths={1}
				  isOutsideRange={() => false}
				  showClearDates
				/>
			</div>
		);
	}
}

const mapStateToProps = (state) => ({filters: state.filters});

const mapDispatchToProps = (dispatch) => ({
	sortByAmount: () => {dispatch(sortByAmount())},
	sortByDate: () => {dispatch(sortByDate())},
	setStartDate: (startDate) => {dispatch(setStartDate(startDate))},
	setEndDate: (endDate) => {dispatch(setEndDate(endDate))},
	setTextFilter: (text) => {dispatch(setTextFilter(text))}
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);