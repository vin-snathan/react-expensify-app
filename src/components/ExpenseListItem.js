import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({description, amount, createdAt, id}) => {

	return (
		<div>
			<Link to={`/edit/${id}`}><h4> {description} </h4></Link>
			<p> 
				{numeral(amount/100).format('$0,0.00')} 
				- 
				{moment(createdAt).format('MMMM Do, YYYY')} 
			</p>
		</div>
	);
};

export default ExpenseListItem;