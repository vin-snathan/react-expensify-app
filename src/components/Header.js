import React from 'react';
import {NavLink} from 'react-router-dom';

const Header = (props) => (
	<div>
		<h1> Expensify App </h1>
		<NavLink activeClassName='selected' exact to='/'>Expense Dashboard</NavLink>
		<NavLink activeClassName='selected' to='/create'>Create Expense</NavLink>
		<NavLink activeClassName='selected' to='/help'>Help</NavLink>
	</div>
);

export default Header;