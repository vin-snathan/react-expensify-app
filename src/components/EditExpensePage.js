import React, {Component} from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense, removeExpense} from '../actions/expenses';

export class EditExpensePage extends Component {
	constructor(props) {
		super(props);
		this.onSubmit = this.onSubmit.bind(this);
		this.handleRemove = this.handleRemove.bind(this);
	}

	onSubmit(expense) {
		this.props.editExpense(this.props.expense, expense); 
		this.props.history.push('/');
	}
	
	handleRemove() {
		this.props.removeExpense(this.props.expense);
		this.props.history.push('/');
	}

	render() {
		return (
			<div>
				<h1>Edit Expense Page</h1>
				<ExpenseForm {...this.props.expense} onSubmit={this.onSubmit} />
				<button onClick={this.handleRemove}>Remove</button>
			</div>
		)
	}

}

const mapStateToProps = (state, props) => ({expense: state.expenses.find(({id}) => id === props.match.params.id)})

const mapDispatchToProps = (dispatch) => ({
	editExpense: (id, expense) => dispatch(editExpense(id, expense)),
	removeExpense: (data) => dispatch(removeExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);