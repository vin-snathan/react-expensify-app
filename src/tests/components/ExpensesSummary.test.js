import React from 'react';
import {shallow} from 'enzyme';
import toJSON from 'enzyme-to-json';
import {ExpensesSummary} from '../../components/ExpensesSummary';

test('Should correctly render ExpenseSummary with 1 expense', () => {
	const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesSummary={234} />);
	expect(wrapper).toMatchSnapshot();
})

test('Should correctly render ExpenseSummary with multiple expenses', () => {
	const wrapper = shallow(<ExpensesSummary expensesCount={5} expensesSummary={23343434} />);
	expect(wrapper).toMatchSnapshot();
})