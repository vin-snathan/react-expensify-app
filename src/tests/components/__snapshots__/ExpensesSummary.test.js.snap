// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should correctly render ExpenseSummary with 1 expense 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    expensesCount={1}
    expensesSummary={234}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <h2>
        Viewing 
         
        expenses
         totalling 
        $0.00
</h2>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "Viewing ",
          undefined,
          " ",
          "expenses",
          " totalling ",
          "$0.00",
        ],
      },
      "ref": null,
      "rendered": Array [
        "Viewing ",
        undefined,
        " ",
        "expenses",
        " totalling ",
        "$0.00",
      ],
      "type": "h2",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <h2>
          Viewing 
           
          expenses
           totalling 
          $0.00
</h2>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Viewing ",
            undefined,
            " ",
            "expenses",
            " totalling ",
            "$0.00",
          ],
        },
        "ref": null,
        "rendered": Array [
          "Viewing ",
          undefined,
          " ",
          "expenses",
          " totalling ",
          "$0.00",
        ],
        "type": "h2",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Should correctly render ExpenseSummary with multiple expenses 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    expensesCount={5}
    expensesSummary={23343434}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <h2>
        Viewing 
         
        expenses
         totalling 
        $0.00
</h2>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "Viewing ",
          undefined,
          " ",
          "expenses",
          " totalling ",
          "$0.00",
        ],
      },
      "ref": null,
      "rendered": Array [
        "Viewing ",
        undefined,
        " ",
        "expenses",
        " totalling ",
        "$0.00",
      ],
      "type": "h2",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <h2>
          Viewing 
           
          expenses
           totalling 
          $0.00
</h2>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Viewing ",
            undefined,
            " ",
            "expenses",
            " totalling ",
            "$0.00",
          ],
        },
        "ref": null,
        "rendered": Array [
          "Viewing ",
          undefined,
          " ",
          "expenses",
          " totalling ",
          "$0.00",
        ],
        "type": "h2",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
